plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt

```diff
@@ -305,12 +286,30 @@ data class ClassTypeEmbedding(val className: ScopedKotlinName) : TypeEmbedding {
         flatMapUniqueFields { _, field -> field.accessInvariantsForParameter() }
 
     // Note: this function will replace accessInvariants when nested unfold will be implemented
-    override fun predicateAccessInvariants(): List<TypeInvariantEmbedding> = listOf(PredicateAccessTypeInvariantEmbedding(name))
-
+    override fun predicateAccessInvariant() =
+        PredicateAccessTypeInvariantEmbedding(name, PermExp.WildcardPerm())
+
+    // Returns the list of classes in a hierarchy that need to be unfolded in order to access the given field
+    fun hierarchyUnfoldPath(fieldName: MangledName): List<ClassTypeEmbedding> {
```

Comment:
I wonder whether this could be written as a generator to produce an iterator... May be overkill, though